projects/distributed-tracing/src/shared/dashboard/widgets/table/services/table-widget-columns.service.ts

```diff
@@ -0,0 +1,87 @@
+import { Injectable } from '@angular/core';
+import { CoreTableCellRendererType, FilterBuilderLookupService, TableCellAlignmentType } from '@hypertrace/components';
+import { Observable } from 'rxjs';
+import { map } from 'rxjs/operators';
+import {
+  AttributeMetadata,
+  AttributeMetadataType,
+  toFilterAttributeType
+} from '../../../../graphql/model/metadata/attribute-metadata';
+import { SpecificationBuilder } from '../../../../graphql/request/builders/specification/specification-builder';
+import { MetadataService } from '../../../../services/metadata/metadata.service';
+import { SpecificationBackedTableColumnDef } from '../table-widget-column.model';
+
+@Injectable({ providedIn: 'root' })
+export class TableWidgetColumnsService {
+  public constructor(
+    private readonly metadataService: MetadataService,
+    private readonly filterBuilderLookupService: FilterBuilderLookupService
+  ) {}
+
+  public fetchColumn(
+    scope: string,
+    existingColumns: SpecificationBackedTableColumnDef[]
+  ): Observable<SpecificationBackedTableColumnDef[]> {
+    return this.metadataService.getSelectionAttributes(scope).pipe(
+      map(attributes => [
+        ...existingColumns,
+        ...attributes
+          .filter(attribute => !this.isExplicitlyDeclaredAttribute(attribute, existingColumns))
+          .filter(attribute => !this.isAggregationOnlyAttribute(attribute))
+          .flatMap(attribute => this.mapAttributeToColumnConfig(attribute))
```

Comment:
flatMap means mapAttributeToColumnConfig should return an array

Reply:
Ah, this was something that I started changing and forgot to go back and remove. Fixed. Thanks.